support/man/strsep.3
.\"
.\" Copyright (c) 1999,2000 WU-FTPD Development Group. 
.\" All rights reserved.
.\" 
.\" Portions Copyright (c) 1980, 1985, 1988, 1989, 1990, 1991, 1993, 1994 
.\" The Regents of the University of California.  Portions Copyright (c) 
.\" 1993, 1994 Washington University in Saint Louis.  Portions Copyright 
.\" (c) 1996, 1998 Berkeley Software Design, Inc.  Portions Copyright (c) 
.\" 1998 Sendmail, Inc.  Portions Copyright (c) 1983, 1995, 1996, 1997 Eric 
.\" P. Allman.  Portions Copyright (c) 1989 Massachusetts Institute of 
.\" Technology.  Portions Copyright (c) 1997 Stan Barber.  Portions 
.\" Copyright (C) 1991, 1992, 1993, 1994, 1995, 1996, 1997 Free Software 
.\" Foundation, Inc.  Portions Copyright (c) 1997 Kent Landfield. 
.\"
.\" Use and distribution of this software and its source code are governed 
.\" by the terms and conditions of the WU-FTPD Software License ("LICENSE"). 
.\"
.\"     $Id: strsep.3,v 1.4 2000/07/01 18:40:49 wuftpd Exp $
.\"
.TH STRSEP 3 "May 15, 1990"
.UC 7
.SH NAME
strsep \- separate strings
.SH SYNOPSIS
.nf
.ft B
#include <string.h>

char *
strsep(char **stringp, char *delim);
.ft R
.fi
.SH DESCRIPTION
.B Strsep
locates in the null-terminated string at
.I *stringp
the first occurence of any character in
.I delim
and replaces this with a '\e0',
records the location of the immediate following character in
.IR *stringp ,
then returns the original value of
.IR *stringp .
If no delimiter characters are found,
.B strsep
sets
.I *stringp
to NULL;
if
.I *stringp
is initially NULL,
.B strsep
returns NULL.
.SH EXAMPLES
The following uses
.I strsep
to parse strings containing runs of white space,
making up an argument vector:
.sp
.nf
.RS
char inputstring[100];
char **argv[51], **ap = argv, *p, *val;
.I "/* set up inputstring */"
for (p = inputstring; p != NULL; ) {
    while ((val = strsep(&p, " \et")) != NULL && *val == '\e0');
    *ap++ = val;
}
*ap = 0;
.RE


CHANGES
 
  Copyright (c) 1999,2000,2001 WU-FTPD Development Group.  
  All rights reserved.
  
  Portions Copyright (c) 1980, 1985, 1988, 1989, 1990, 1991, 1993, 1994
    The Regents of the University of California.
  Portions Copyright (c) 1993, 1994 Washington University in Saint Louis.
  Portions Copyright (c) 1996, 1998 Berkeley Software Design, Inc.
  Portions Copyright (c) 1989 Massachusetts Institute of Technology.
  Portions Copyright (c) 1998 Sendmail, Inc.
  Portions Copyright (c) 1983, 1995, 1996, 1997 Eric P.  Allman.
  Portions Copyright (c) 1997 Stan Barber.
  Portions Copyright (c) 1997 Kent Landfield.
  Portions Copyright (c) 1991, 1992, 1993, 1994, 1995, 1996, 1997
    Free Software Foundation, Inc.  
 
  Use and distribution of this software and its source code are governed 
  by the terms and conditions of the WU-FTPD Software License ("LICENSE").
 
  If you did not receive a copy of the license, it may be obtained online
  at http://www.wu-ftpd.org/license.html.
 
  $Id: CHANGES,v 1.44.2.1 2001/11/29 17:25:33 wuftpd Exp $


Changes in 2.6.2: Released 29 Nov, 2001

 o  Added checks for missing "]" and "}" in filename globs, this completes
    the file globbing heap corruption vulnerability fix.

 o  Added checks to the globbing code for overflow of restbuf, and additional
    globerr setting and checking to speed up return on error.

 o  Changed the globbing code to use qsort, much faster when sorting a large
    number of strings.

 o  Handle ftpglob() returning a vector containing just a NULL string, fixes
    problems caused by CWD ~{

 o  Somehow the fix for pasv-allow didn't actually make it into 2.6.1

 o  Provide a compile-time option to revert NLST to showing directories.

 o  Fix missing format strings in debugging code.

Changes in 2.6.1: Released 2 Jul, 2000

 o  Fix security leaks that could result in a root shell compromise.

 o  Fix memory leaks in internal ls (this feature still needs more testing;
    you should probably not use it on high-traffic production servers yet.)

 o  Fix up the port-allow command in ftpaccess.

 o  Merge in the virtual passwd/virtual shadow features of BeroFTPD.

 o  Some fixes to the configure script.

 o  SITE MINFO was missed in 2.6.0 when disabling SITE NEWER.

 o  Fix documentation of data-limit.

Changes in 2.6.0: Released 18 Oct, 1999

 o  On sigpipe, always log a lost connection.

 o  Added a log message on attempts to download files marked unretrievable.

 o  The SITE NEWER feature has been disabled.  A compile-time option has been
    added to re-enable it.  See config.h.noac for more information on this.

 o  With restricted-uid/gid, CWD to a non-existant directory would display the
    full pathname rather than just relative to the user's home.  Actually, the
    fix catches most cases where this could occur, not just the CWD verb.

 o  Fixed a bug in the restricted-uid/gid feature which could allow access
    outside the user's home directory in some cases.

 o  Bumped MAXHST (max. hosts allowed on a line) for ftphosts from 10 to 12.
    Fixed a bug related to this which can cause the server to crash checking
    host access.

 o  The internal ls (see below) was judged to be unready.  It has been disabled
    by default but can be enabled with a compile-time option for those who wish
    to attempt to debug it (be warned, it has a lot of problems).

 o  Split the "bad shell or user not in ftpusers" syslog message into two
    messages to prevent confusion.

 o  Filename globs for LIST, NLST and SITE EXEC, as well as a few internal
    uses, are cleaned up before processing.  For example: */./../* becomes
    just *.  This prevents certain memory starvation DoS attacks.

 o  Corrections for RFC compliance can break some clients.  If possible, the
    broken client should be updated, but a compile-time option has been
    added.  See the config.h.noac for more information on this.

 o  Created doc/HOWTO directory and moved VIRTUAL.FTP.SUPPORT and 
    upload.configuration.HOWTO there.

 o  Add a README.AUTOCONF file describing the autoconf build in detail.

 o  UC, Berkeley, has removed the requirement that all advertising material
    must include credit to them.  Removed the clause from the LICENSE and
    the historical licenses in the COPYRIGHT file.

 o  Added the email-on-upload feature from BeroFTPD.  See the ftpaccess man
    page for defaults on these added ftpaccess clauses:

        mailserver <hostname>
        incmail <emailaddress>
        mailfrom <emailaddress>
        virtual <address> incmail <emailaddress>
        virtual <address> mailfrom <emailaddress>
        defaultserver incmail <emailaddress>
        defaultserver mailfrom <emailaddress>

 o  Redhat added the -I option to disable RFC931 (AUTH/ident).  Added to
    the baseline so Redhat users don't see a loss of a feature.  Setting
    the timeout for rfc931 to zero will do the same thing in the ftpaccess
    file.

 o  The test for whether restricted-uid/restricted-gid applied should have
    been done before the chroot so it used the system /etc/passwd and
    /etc/group files.

 o  CDUP when you were already at the home directory, would complain about
    you being restricted (if you were).  Instead it should give a positive
    reply, and do nothing.  This makes it behave more like CDUP when you're
    not restricted to your home directory.

 o  deny-uid and deny-gid were being tested for anonymous users.  Bad move,
    it's too easy to forget to allow them.  Use 'defaultserver private' to
    keep anonymous users away.

 o  Correct the operation of the NLST command.  Finally.  mget should now
    work as users expect it to.

 o  Prevent buffer overruns when processing message files.

 o  Correct a reference through a NULL pointer when doing S/Key
    authentication and the user is not in the passwd file.

 o  Check the return code from select() when setting up a data connection.

support/strsep.c
/****************************************************************************  
 
  Copyright (c) 1999,2000 WU-FTPD Development Group.  
  All rights reserved.
  
  Portions Copyright (c) 1980, 1985, 1988, 1989, 1990, 1991, 1993, 1994
    The Regents of the University of California.
  Portions Copyright (c) 1993, 1994 Washington University in Saint Louis.
  Portions Copyright (c) 1996, 1998 Berkeley Software Design, Inc.
  Portions Copyright (c) 1989 Massachusetts Institute of Technology.
  Portions Copyright (c) 1998 Sendmail, Inc.
  Portions Copyright (c) 1983, 1995, 1996, 1997 Eric P.  Allman.
  Portions Copyright (c) 1997 by Stan Barber.
  Portions Copyright (c) 1997 by Kent Landfield.
  Portions Copyright (c) 1991, 1992, 1993, 1994, 1995, 1996, 1997
    Free Software Foundation, Inc.  
 
  Use and distribution of this software and its source code are governed 
  by the terms and conditions of the WU-FTPD Software License ("LICENSE").
 
  If you did not receive a copy of the license, it may be obtained online
  at http://www.wu-ftpd.org/license.html.
 
  $Id: strsep.c,v 1.6 2000/07/01 18:36:29 wuftpd Exp $
 
****************************************************************************/
#include <string.h>
#include <stdio.h>

/*
 * Get next token from string *stringp, where tokens are nonempty
 * strings separated by characters from delim.
 *
 * Writes NULLs into the string at *stringp to end tokens.
 * delim need not remain constant from call to call.
 * On return, *stringp points past the last NUL written (if there might
 * be further tokens), or is NULL (if there are definitely no more tokens).
 *
 * If *stringp is NULL, strsep returns NULL.
 */
char *strsep(register char **stringp, register const char *delim)
{
    register char *s;
    register const char *spanp;
    register int c, sc;
    char *tok;

    if ((s = *stringp) == NULL)
	return (NULL);
    for (tok = s;;) {
	c = *s++;
	spanp = delim;
	do {
	    if ((sc = *spanp++) == c) {
		if (c == 0)
		    s = NULL;
		else
		    s[-1] = 0;
		*stringp = s;
		return (tok);
	    }
	} while (sc != 0);
    }
    /* NOTREACHED */
}
